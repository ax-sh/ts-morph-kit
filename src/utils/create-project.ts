import { Project, type ProjectOptions } from "ts-morph";

export function createProject(options?: ProjectOptions) {
  const project = new Project({
    compilerOptions: {},
    ...options,
  });
  return project;
}

export function openAsSourceFile(filePath: string) {
  const project = createProject();
  const sourceFile = project.addSourceFileAtPath(filePath);

  return sourceFile;
}
