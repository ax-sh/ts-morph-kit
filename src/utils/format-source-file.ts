import type { SourceFile } from "ts-morph"
import { ts } from "ts-morph"
// import SemicolonPreference = ts.SemicolonPreference;

export function formatSourceFile(sourceFile: SourceFile) {
  sourceFile.formatText({
    indentSize: 2,
    semicolons: ts.SemicolonPreference.Insert,
    convertTabsToSpaces: true,
  })
  return sourceFile
}

export function formatSourceFileToString(sourceFile: SourceFile) {
  return formatSourceFile(sourceFile).getFullText()
}
