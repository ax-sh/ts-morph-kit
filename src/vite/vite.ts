import { findDefaultExport } from "../utils/find-default-export.ts";
import { CallExpression, type SourceFile, SyntaxKind } from "ts-morph";

export class NotAFunctionCallError extends Error {
  constructor() {
    super("The 'export default' is not a function call.");
    this.name = "NotAFunctionCallError";
  }
}

export class InvalidCalleeError extends Error {
  constructor(expectedFuncName: string) {
    super(`The 'export default' does not call '${expectedFuncName}'.`);
    this.name = "InvalidCalleeError";
  }
}

export class MissingArgumentsError extends Error {
  constructor(funcName: string) {
    super(`The '${funcName}' call has no arguments.`);
    this.name = "MissingArgumentsError";
  }
}

export class InvalidArgumentTypeError extends Error {
  constructor(funcName: string) {
    super(
      `The '${funcName}' call does not contain an object literal as its argument.`,
    );
    this.name = "InvalidArgumentTypeError";
  }
}

export function getDefaultViteConfig(sourceFile: SourceFile) {
  const exportedExpression = findDefaultExport(sourceFile);
  // Ensure the exported expression is a CallExpression
  if (!CallExpression.isCallExpression(exportedExpression)) {
    throw new NotAFunctionCallError();
  }
  // Check if the callee is `defineConfig`
  const callee = exportedExpression.getExpression();
  const funcName = "defineConfig";
  if (
    callee.getKind() !== SyntaxKind.Identifier ||
    callee.getText() !== funcName
  ) {
    throw new InvalidCalleeError(funcName);
  }
  // Get the arguments of the `defineConfig` call
  const args = exportedExpression.getArguments();
  if (args.length === 0) {
    throw new MissingArgumentsError(funcName);
  }

  const [configObject] = args;
  if (
    !configObject ||
    !configObject.isKind(SyntaxKind.ObjectLiteralExpression)
  ) {
    throw new InvalidArgumentTypeError(funcName);
  }
  return configObject;
}
