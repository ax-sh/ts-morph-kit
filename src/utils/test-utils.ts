import path from "node:path";
import { ScriptKind } from "ts-morph";
import { createProject } from "../core/create-project.ts";

export async function createMemorySourceFile(tempFile: string, input: string) {
  const project = createProject({ useInMemoryFileSystem: true });
  const sourceFile = project.createSourceFile(tempFile, input, {
    // Note: .js and .mjs can still be valid for TS projects.
    // We can't infer TypeScript from config.tsx.
    scriptKind:
      path.extname(tempFile) === ".ts" ? ScriptKind.TS : ScriptKind.JS,
  });

  return sourceFile;
}

export async function createTestSourceFile(code: string) {
  const resolvedPath = "test.ts";
  const sourceFile = await createMemorySourceFile(resolvedPath, code);
  return sourceFile;
}
