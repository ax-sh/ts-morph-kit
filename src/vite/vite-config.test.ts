import { openAsSourceFile } from "../core/create-project.ts";
import { formatSourceFile, formatSourceFileToString } from "../utils/format-source-file.ts";
import { createTestSourceFile } from "../utils/test-utils.ts";
import { addBaseProperty, addVitePlugins } from "./vite.ts";

const code = `
        import { defineConfig } from 'vite';
        // https://vite.dev/config/
        export default defineConfig({
        plugins:[foo()],
        base:"test"
        });
  `;
vi.mock("../core/create-project.ts", async () => {
  const actual = await vi.importActual("../core/create-project.ts");
  return {
    ...actual,
    // createProject: vi.fn(),
    openAsSourceFile: vi.fn(),
  };
});

describe("openAsSourceFile mock", () => {
  // Before your tests, make sure to resolve the mock
  beforeEach(async () => {
    vi.mocked(openAsSourceFile).mockReturnValue(
      await createTestSourceFile(code),
    );
  });
  it("should mock the function and return a custom value", async () => {
    const result = openAsSourceFile("test.ts");
    result.formatText();

    // const mockResult = { content: "Mocked file content" };
    // expect(result).toEqual(mockResult);
    expect(openAsSourceFile).toHaveBeenCalledWith("test.ts");
    expect(openAsSourceFile).toHaveBeenCalledTimes(1);
  });
  it("[addBaseProperty] modify base without duplicates", async () => {
    const sf = openAsSourceFile("test.ts");
    const modified = addBaseProperty(sf, "voo");
    const formatted = formatSourceFileToString(modified.getSourceFile());
    // Assert that the base property is added correctly
    expect(formatted).toContain('base: "voo"');

    // Assert that the base property is not duplicated
    const basePropertyCount = (formatted.match(/base:/g) || []).length;
    expect(basePropertyCount).toBe(1);
  });
  it("should test formatting logic to source file", async () => {
    const sourceFile = openAsSourceFile("test.ts");
    expect(openAsSourceFile).toHaveBeenCalledWith("test.ts");
    const formatted = formatSourceFile(sourceFile).getText();
    expect(formatted).toMatchSnapshot();
  });

  it.todo("should modify plugins without duplicates", async () => {
    const sf = openAsSourceFile("test.ts");
    const modified = addVitePlugins(sf, ["dff()", "dff", "dff"]);
    sf.formatText()
    const before = sf.getText()
    modified.formatText()
    const formatted = formatSourceFileToString(modified.getSourceFile());

    expect(before).toBe(formatted);
  });
});
