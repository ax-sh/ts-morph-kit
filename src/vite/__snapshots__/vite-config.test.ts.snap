// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`openAsSourceFile mock > should test formatting logic to source file 1`] = `
"import { defineConfig } from 'vite';
// https://vite.dev/config/
export default defineConfig({
  plugins: [foo()],
  base: "test"
});
"
`;
